libraries/chain/combined_database.cpp

```diff
@@ -0,0 +1,470 @@
+                    "rocksdb");
+         db.modify(db.get<kv_db_config_object>(), [](auto& cfg) { cfg.backing_store = backing_store_type::ROCKSDB; });
+      }
+      if (db.get<kv_db_config_object>().backing_store == backing_store_type::ROCKSDB)
+         ilog("using rocksdb for backing store");
+      else
+         ilog("using chainbase for backing store");
+   }
+
+   void combined_database::set_revision(uint64_t revision) {
+      try {
+         try {
+            db.set_revision(revision);
+#warning TODO: Chain_kv needs to fix kv_undo_stack's revision after restart
+            // kv_undo_stack.set_revision(revision, false);
+         }
+         FC_LOG_AND_RETHROW()
+      }
+      CATCH_AND_EXIT_DB_FAILURE()
+   }
+
+   void combined_database::undo() {
+      try {
+         try {
+            db.undo();
+            kv_undo_stack.undo(false);
+         }
+         FC_LOG_AND_RETHROW()
+      }
+      CATCH_AND_EXIT_DB_FAILURE()
+   }
+
+   void combined_database::commit(int64_t revision) {
+      try {
+         try {
+            db.commit(revision);
+            kv_undo_stack.commit(revision);
+         }
+         FC_LOG_AND_RETHROW()
+      }
+      CATCH_AND_EXIT_DB_FAILURE()
+   }
+
+   void combined_database::flush() {
+      try {
+         try {
+            kv_undo_stack.write_state();
+            kv_database.flush(true, true);
+         }
+         FC_LOG_AND_RETHROW()
+      }
+      CATCH_AND_EXIT_DB_FAILURE()
+   }
+
+   std::unique_ptr<kv_context> combined_database::create_kv_context(name receiver, kv_resource_manager resource_manager,
+                                                                    const kv_database_config& limits) {
+      switch (backing_store) {
+         case backing_store_type::ROCKSDB:
+            return create_kv_rocksdb_context(kv_database, kv_undo_stack, receiver, resource_manager, limits);
+         case backing_store_type::NATIVE: return create_kv_chainbase_context(db, receiver, resource_manager, limits);
+         default: EOS_ASSERT(false, action_validate_exception, "Unknown backing store.");
+      }
+   }
+
+   void combined_database::add_to_snapshot(
+         const eosio::chain::snapshot_writer_ptr& snapshot, const eosio::chain::block_state& head,
+         const eosio::chain::authorization_manager&                    authorization,
+         const eosio::chain::resource_limits::resource_limits_manager& resource_limits) const {
+
+      snapshot->write_section<chain_snapshot_header>(
+            [this](auto& section) { section.add_row(chain_snapshot_header(), db); });
+
+      snapshot->write_section<block_state>(
+            [this, &head](auto& section) { section.template add_row<block_header_state>(head, db); });
+
+      eosio::chain::controller_index_set::walk_indices([this, &snapshot](auto utils) {
+         using value_t = typename decltype(utils)::index_t::value_type;
```

Comment:
`utils` in this context will always copy.  Is that what is intended?  If not change to `auto&` and `std::decay_t<decltype(utils)>`.

Reply:
Because all four occurrences of similar constructs pass `utils` by value, I assume that was the author's intent.